pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go

```diff
@@ -222,136 +306,220 @@ func (m *rangefeedMuxer) startNodeMuxRangeFeed(
+					active.catchupRes = nil
 				}
-				continue
+				// Note that this timestamp means that all rows in the span with
+				// writes at or before the timestamp have now been seen. The
+				// Timestamp field in the request is exclusive, meaning if we send
+				// the request with exactly the ResolveTS, we'll see only rows after
+				// that timestamp.
+				active.startAfter.Forward(t.ResolvedTS)
 			}
+		case *kvpb.RangeFeedSSTable:
+		case *kvpb.RangeFeedError:
+			log.VErrEventf(ctx, 2, "RangeFeedError: %s", t.Error.GoError())
+			if active.catchupRes != nil {
+				m.ds.metrics.RangefeedErrorCatchup.Inc(1)
+			}
+			conn.streams.Delete(event.StreamID)
+			if _, err := handleRangefeedError(ctx, t.Error.GoError()); err != nil {
+				// If this is an error we cannot recover from, terminate the rangefeed.
+				return err
+			}
+			if err := m.restartActiveRangeFeed(ctx, active, t.Error.GoError()); err != nil {
+				return err
+			}
+			continue
+		}
 
-			select {
-			case <-ctx.Done():
-				return ctx.Err()
-			case producer.eventCh <- &e.RangeFeedEvent:
-			case <-producer.muxClientCtx.Done():
-				if log.V(1) {
-					log.Infof(ctx, "received stray event stream %d: %v", e.StreamID, e)
+		active.onRangeEvent(nodeID, event.RangeID, &event.RangeFeedEvent)
+		msg := RangeFeedMessage{RangeFeedEvent: &event.RangeFeedEvent, RegisteredSpan: active.Span}
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case m.eventCh <- msg:
+		}
+
+		// Piggyback on this loop to check if any of the active ranges
+		// on this node appear to be stuck.
+		// NB: this does not notify the server in any way.  We may have to add
+		// a more complex protocol -- or better yet, figure out why ranges
+		// get stuck in the first place.
+		if timeutil.Now().Before(nextStuckCheck) {
+			if threshold := stuckThreshold(); threshold > 0 {
+				if err := conn.eachStream(func(id int64, a *activeMuxRangeFeed) error {
+					if !a.startAfter.IsEmpty() && timeutil.Since(a.startAfter.GoTime()) > stuckThreshold() {
+						conn.streams.Delete(id)
+						return m.restartActiveRangeFeed(ctx, a, errRestartStuckRange)
+					}
+					return nil
+				}); err != nil {
+					return err
 				}
 			}
+			nextStuckCheck = timeutil.Now().Add(stuckCheckFreq())
 		}
 	}
 }
 
-// terminationContext (inspired by context.Context) describes
-// termination information.
-type terminationContext interface {
-	Done() <-chan struct{}
-	Err() error
-}
+func (m *rangefeedMuxer) restartActiveRangeFeed(
+	ctx context.Context, active *activeMuxRangeFeed, reason error,
+) error {
+	if log.V(1) {
+		log.Infof(ctx, "RangeFeed %s@%s disconnected with last checkpoint %s ago: %v",
+			active.Span, active.StartAfter, timeutil.Since(active.Resolved.GoTime()), reason)
+	}
+	active.setLastError(reason)
 
-// termCtx implements terminationContext, and allows error to be set.
-type termCtx struct {
-	sync.Once
-	done chan struct{}
-	err  error
-}
+	active.release()
+	active.token.Evict(ctx)
```

Comment:
Same comment as above, we should only evict this when we need to since it affects foreground tail latencies.